Mathlib/Computability/TuringDegrees.lean

```diff
@@ -0,0 +1,355 @@
+  AntisymmRel RecursiveIn f g
+
+/--
+Custom infix notation for `turing_equivalent`.
+-/
+infix:50 " ≡ᵀ " => TuringEquivalent
+
+/--
+If a function is partial recursive, then it is recursive in every partial function.
+-/
+lemma partrec_implies_recursive_in_everything
+  (f : ℕ →. ℕ) : Nat.Partrec f → (∀ g, RecursiveIn f g) := by
+    intro pF
+    intro g
+    induction pF
+    case zero =>
+      apply RecursiveIn.zero
+    case succ =>
+      apply RecursiveIn.succ
+    case left =>
+      apply RecursiveIn.left
+    case right =>
+      apply RecursiveIn.right
+    case pair _ _ _ _ ih1 ih2 =>
+      apply RecursiveIn.pair ih1 ih2
+    case comp _ _ _ _ ih1 ih2 =>
+      apply RecursiveIn.comp ih1 ih2
+    case prec _ _ _ _ ih1 ih2 =>
+      apply RecursiveIn.prec ih1 ih2
+    case rfind _ _ ih =>
+      apply RecursiveIn.rfind ih
+
+/--
+If a function is recursive in the constant zero function,
+then it is partial recursive.
+-/
+lemma partrec_in_zero_implies_partrec
+(f : ℕ →. ℕ) : RecursiveIn f (fun _ => Part.some 0) → Nat.Partrec f := by
+  intro fRecInZero
+  generalize h : (fun _ => Part.some 0) = fp at *
+  induction fRecInZero
+  case zero =>
+    apply Nat.Partrec.zero
+  case succ =>
+    apply Nat.Partrec.succ
+  case left =>
+    apply Nat.Partrec.left
+  case right =>
+    apply Nat.Partrec.right
+  case oracle g =>
+    rw [← h]
+    apply Nat.Partrec.zero
+  case pair _ _ _ _ ih1 ih2 =>
+    apply Nat.Partrec.pair
+    · apply ih1
+      rw [← h]
+    · apply ih2
+      rw [← h]
+  case comp _ _ _ _ ih1 ih2 =>
+    apply Nat.Partrec.comp
+    · apply ih1
+      rw [← h]
+    · apply ih2
+      rw [← h]
+  case prec _ _ _ _ ih1 ih2 =>
+    apply Nat.Partrec.prec
+    · apply ih1
+      rw [← h]
+    · apply ih2
+      rw [← h]
+  case rfind _ _ ih =>
+    apply Nat.Partrec.rfind
+    apply ih
+    rw [← h]
+
+/--
+A partial function `f` is partial recursive if and only if it is recursive in
+every partial function `g`.
+-/
+theorem partrec_iff_partrec_in_everything
+  (f : ℕ →. ℕ) : Nat.Partrec f ↔ (∀ g, RecursiveIn f g) := by
+  constructor
+  · exact partrec_implies_recursive_in_everything f
+  · intro H
+    have lem : RecursiveIn f (fun _ => Part.some 0) := H (fun _ => Part.some 0)
+    have lem : Nat.Partrec f := partrec_in_zero_implies_partrec f lem
+    exact lem
+
+/--
+Proof that turing reducibility is reflexive.
+-/
+theorem RecursiveIn.refl (f : ℕ →. ℕ) : RecursiveIn f f :=
+  RecursiveIn.oracle
+
+/--
+Proof that `turing_equivalent` is reflexive.
+-/
+@[refl]
+theorem TuringEquivalent.refl (f : ℕ →. ℕ) : f ≡ᵀ f :=
+  ⟨RecursiveIn.refl f, RecursiveIn.refl f⟩
+
+/--
+Proof that `turing_equivalent` is symmetric.
+-/
+@[symm]
+theorem TuringEquivalent.symm {f g : ℕ →. ℕ} (h : f ≡ᵀ g) : g ≡ᵀ f :=
+  ⟨h.2, h.1⟩
+
+/--
+Proof that turing reducibility is transitive.
+-/
+@[trans]
+theorem RecursiveIn.trans {f g h : ℕ →. ℕ} :
+  RecursiveIn f g → RecursiveIn g h → RecursiveIn f h := by
+    intro hg hh
```

Comment:
In general instead of writing function types and then using `intro` you can introduce values directly as arguments - often this is cleaner.
```suggestion
@[trans]
theorem RecursiveIn.trans {f g h : ℕ →. ℕ} (hg : RecursiveIn f g) (hh : RecursiveIn g h) :
    RecursiveIn f h := by
```